sample/trace_key.go

```diff
@@ -112,3 +108,92 @@ outer:
 
 	return key.String(), fieldCount
 }
+
+// distinctValue keeps track of distinct values for a set of fields.
+// It stores the unique values as strings.
+type distinctValue struct {
+	buf    []byte
+	fields []string
+	values []map[uint64]string
+
+	// totalUniqueCount keeps track of how many unique values we've seen so far for a trace key.
+	totalUniqueCount int
+	// maxDistinctValue is the maximum number of distinct values we will store for a trace key.
+	maxDistinctValue int
+}
+
+func newDistinctValue(fields []string, maxDistinctValue int) *distinctValue {
+	d := &distinctValue{
+		buf:              make([]byte, 0, 1024),
+		fields:           fields,
+		values:           make([]map[uint64]string, len(fields)),
+		maxDistinctValue: maxDistinctValue,
+	}
+
+	for i := range d.values {
+		d.values[i] = make(map[uint64]string, maxDistinctValue)
+	}
+
+	return d
+
+}
+
+// Values returns the distinct values for a given field index.
+// It returns a sorted slice of strings containing the unique values for that field.
+func (d *distinctValue) Values(fieldIdx int) []string {
+	if fieldIdx < 0 || fieldIdx >= len(d.values) {
+		return nil
+	}
+
+	// Get the map for the specified field index
+	valueMap := d.values[fieldIdx]
+	if len(valueMap) == 0 {
+		return nil
+	}
+
+	values := slices.Collect(maps.Values(valueMap))
```

Comment:
`slices.Collect` is going to allocate a new slice every time; if you take the pool approach, you could retain and re-use a slice buffer for this instead.